File: src/grid.rs
use std::{arch::x86_64, cell::Cell};

use rand::{Rng, random, rngs};

use crate::grid;

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum CellState {
    Clean,
    SurfaceRust,
    HeavyRust,
    Rotten
}

impl CellState {
    pub fn to_char(&self) -> char {
        match self {
            CellState::Clean => 'O',
            CellState::SurfaceRust => '@',
            CellState::HeavyRust => '#',
            CellState::Rotten => 'X'
        }
    }
}

pub struct Grid {
    pub width: usize,
    pub height: usize,
    pub rust: Vec<Vec<CellState>>,
    pub humidity: Vec<Vec<f32>>,
    pub oxygen: Vec<Vec<f32>>
}

impl Grid {
    pub fn new(width: usize, height: usize) -> Self {
        Grid { width,
            height,
            rust: vec![vec![CellState::Clean; width]; height],
            humidity: vec![vec![0.5; width]; height],
            oxygen: vec![vec![0.5; width]; height]
        }
    }   

    pub fn new_random(width: usize, height: usize) -> Self {
        let mut rng = rand::thread_rng();
        let mut grid = Grid::new(width, height);

        for y in 0..height {
            for x in 0..width {
                grid.humidity[y][x] = rng.gen_range(0.0..1.0);
                grid.oxygen[y][x] = rng.gen_range(0.0..1.0);

                if rng.gen_range(0.0..1.0) < 0.1 {
                    grid.rust[y][x] = CellState::SurfaceRust;
                 }

            }
        } 
        grid
    }

    pub fn display(&self) {
        for row in &self.rust {
            let line: String = row.iter().map(|c| c.to_char()).collect();
            println!("{}", line);
        }
    }

    pub fn get_neighbors(&self, x: usize, y: usize) -> Vec<(usize, usize)> {
        let mut neighbors = Vec::new();
        let x = x as isize;
        let y = y as isize;

        for dy in -1..=1 {
            for dx in -1..=1 {
                if dx == 0 && dy == 0 { continue; }
                let nx = x + dx;
                let ny = y + dy;
                if nx >= 0 && ny >= 0 && nx < self.width as isize && ny < self.height as isize {
                    neighbors.push((nx as usize, ny as usize));
                }
            }
        }

        neighbors
    }

}

File: python/grid.py
import numpy as np
import json
import os

CLEAN        = 0
SURFACE_RUST = 1
HEAVY_RUST   = 2
ROTTEN       = 3


STATE_CHAR = {
    CLEAN:        '0',
    SURFACE_RUST: '1',
    HEAVY_RUST:   '2',
    ROTTEN:       '3'
}

STATE_STR = {
    CLEAN:        'clean',
    SURFACE_RUST: 'surface',
    HEAVY_RUST:   'heavy',
    ROTTEN:       'rotten'
}

def new_grid(width, height, humidity=0.5, oxygen=0.5, rust_seed=0.1):
    rust = np.zeros((height, width), dtype=np.int32)
    hum  = np.full((height, width), humidity, dtype=np.float32)
    oxy  = np.full((height, width), oxygen,   dtype=np.float32)

    mask = np.random.random((height, width)) < rust_seed
    rust[mask] = SURFACE_RUST

    return rust, hum, oxy

def new_grid_random(width, height, rust_seed=0.1):
    rust = np.zeros((height, width), dtype=np.int32)
    hum  = np.random.random((height, width)).astype(np.float32)
    oxy  = np.random.random((height, width)).astype(np.float32)

    mask = np.random.random((height, width)) < rust_seed
    rust[mask] = SURFACE_RUST

    return rust, hum, oxy

def display(rust):
    for row in rust:
        print(' '.join(STATE_CHAR[c] for c in row))


def save_step(rust, hum, oxy, step, output_dir):
    os.makedirs(output_dir, exist_ok=True)
    data = {
        'step':     step,
        'width':    rust.shape[1],
        'height':   rust.shape[0],
        'grid':     [[STATE_STR[c] for c in row] for row in rust.tolist()],
        'humidity': hum.tolist(),
        'oxygen':   oxy.tolist(),
    }
    path = os.path.join(output_dir, f'step_{step:04d}.json')
    with open(path, 'w') as f:
        json.dump(data, f, indent=2)

def save_summary(total_steps, width, height, output_dir):
    data = {
        'total_steps': total_steps,
        'width':       width,
        'height':      height,
        'output_dir':  output_dir,
    }
    path = os.path.join(output_dir, 'summary.json')
    with open(path, 'w') as f:
        json.dump(data, f, indent=2)